File: ruby/food-chain/food_chain.rb
class FoodChain
  VERSION = 2

  SPECIAL = [
    "",
    "It wriggled and jiggled and tickled inside her.\n",
    "How absurd to swallow a bird!\n",
    "Imagine that, to swallow a cat!\n",
    "What a hog, to swallow a dog!\n",
    "Just opened her throat and swallowed a goat!\n",
    "I don't know how she swallowed a cow!\n",
    "She's dead, of course!\n"
  ]
  NAME1 = ["fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"]
  NAME2 = ["fly", "spider that wriggled and jiggled and tickled inside her",
           "bird", "cat", "dog", "goat", "cow", "horse"]
  ENDING = "I don't know why she swallowed the fly. Perhaps she'll die."

  def self.song
    verses(1, 8)
  end
  def self.verses(n, m)
    (n .. m).map{|i| verse(i) }.join("\n\n")
  end
  def self.verse(n)
    text = "I know an old lady who swallowed a #{NAME1[n-1]}.\n#{SPECIAL[n-1]}"
    if n < 8
      text += (n - 1).downto(1).map do |k|
        "She swallowed the #{NAME1[k]} to catch the #{NAME2[k-1]}.\n"
      end.join + ENDING
    end
    text
  end
end

File: exercism/haskell/food-chain/FoodChain.hs
{-# OPTIONS_GHC -XLambdaCase #-}

module FoodChain (song) where

special :: Int -> String
special = \case
  0 -> ""
  1 -> "\nIt wriggled and jiggled and tickled inside her."
  2 -> "\nHow absurd to swallow a bird!"
  3 -> "\nImagine that, to swallow a cat!"
  4 -> "\nWhat a hog, to swallow a dog!"
  5 -> "\nJust opened her throat and swallowed a goat!"
  6 -> "\nI don't know how she swallowed a cow!"
  7 -> "\nShe's dead, of course!"
  n -> error $ "Illegal special id: " ++ show n

animal :: Int -> String
animal = \case
  0 -> "fly"
  1 -> "spider"
  2 -> "bird"
  3 -> "cat"
  4 -> "dog"
  5 -> "goat"
  6 -> "cow"
  7 -> "horse"
  n -> error $ "Illegal animal id: " ++ show n

qualifiedAnimal :: Int -> String
qualifiedAnimal n
  | n == 1 = "spider that wriggled and jiggled and tickled inside her"
  | elem n [0 .. 7] = animal n
  | otherwise = error $ "Illegal qualified animal id: " ++ show n

beginning :: Int -> String
beginning n = concat ["I know an old lady who swallowed a ",
                      animal (n - 1),
                      ".",
                      special (n - 1)]

ending :: String
ending = "I don't know why she swallowed the fly. Perhaps she'll die.\n"

song :: String
song = unlines (map verse [1 .. 8]) ++ "\n"

line :: Int -> String
line k = concat ["She swallowed the ",
                 animal k,
                 " to catch the ",
                 qualifiedAnimal (k - 1),
                 "."]

verse :: Int -> String
verse 8 = beginning 8
verse n = unlines (beginning n : map line (reverse [1 .. (n - 1)])) ++ ending
